openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Merge.cc

```diff
@@ -0,0 +1,682 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+
+/// @file SOP_OpenVDB_Merge.cc
+///
+/// @authors Dan Bailey
+///
+/// @brief Merge OpenVDB grids.
+
+#include <houdini_utils/ParmFactory.h>
+#include <openvdb_houdini/Utils.h>
+#include <openvdb_houdini/PointUtils.h>
+#include <openvdb_houdini/SOP_NodeVDB.h>
+#include <openvdb_houdini/GEO_PrimVDB.h>
+
+#include <openvdb/points/PointDataGrid.h> // points::PointDataGrid
+#include <openvdb/tools/GridTransformer.h> // tools::replaceToMatch()
+#include <openvdb/tools/LevelSetRebuild.h> // tools::doLevelSetRebuild()
+#include <openvdb/tools/Merge.h>
+
+#include <UT/UT_Interrupt.h>
+#include <UT/UT_Version.h>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+
+using namespace openvdb;
+
+namespace hvdb = openvdb_houdini;
+namespace hutil = houdini_utils;
+
+
+class SOP_OpenVDB_Merge: public hvdb::SOP_NodeVDB
+{
+public:
+    SOP_OpenVDB_Merge(OP_Network*, const char* name, OP_Operator*);
+    ~SOP_OpenVDB_Merge() final {}
+
+    static OP_Node* factory(OP_Network*, const char* name, OP_Operator*);
+
+    class Cache: public SOP_VDBCacheOptions
+    {
+    public:
+        fpreal getTime() const { return mTime; }
+    protected:
+        OP_ERROR cookVDBSop(OP_Context&) final;
+    private:
+        fpreal mTime = 0.0;
+    };
+};
+
+
+void
+newSopOperator(OP_OperatorTable* table)
+{
+    if (table == nullptr) return;
+
+    hutil::ParmList parms;
+
+    parms.add(hutil::ParmFactory(PRM_STRING, "group", "Group")
+        .setTooltip("Specify a subset of the input VDBs to be modified.")
+        .setChoiceList(&hutil::PrimGroupMenuInput1)
+        .setDocumentation(
+            "A subset of the input VDBs to be modified"
+            " (see [specifying volumes|/model/volumes#group])"));
+
+    parms.add(hutil::ParmFactory(PRM_STRING, "collation", "Collation")
+        .setDefault("nameclassandtype")
+        .setChoiceListItems(PRM_CHOICELIST_SINGLE, {
+            "nameclassandtype", "Name, VDB Class and Value Type",
+            "classandtype",     "VDB Class and Value Type"
```

Comment:
Since we are doing germancase here I'm not sure why we have the embedded and
I'd prefer
nameclasstype
classtype
Or maybe since these are tokens anyways,
name_class_type
class_type

Reply:
One trivial collation type is "Primitive Number".   Ie, collate the first primitive of all inputs, then the second, etc.  Useful perhaps when people are lazy about naming.

Reply:
Yes, good idea. I took a quick look at doing this and thought it was going to be much more complicated to implement without some significant refactoring then realized I could simply store the primitive number in a hash map keyed by the GEO_PrimVDB::getUniqueId().

After some hesitation around this parameter, I decided it might be better to drop the "VDB Class and Value Type" part for each option and simply make that something that applies to each collation, so it now reads:
* Name
* Primitive Number
* None

If and when I implement something that can merge grids of different value types (ie merging a float grid with a double grid, etc), then this parameter will probably need to change, but this is not a trivial thing to implement anyway so best to keep it simple at this stage I think.